src/raster/types.rs

```diff
@@ -1,41 +1,482 @@
+    }
+
+    /// Get the [`GDALDataType`] size in **bits**.
+    pub fn bits(&self) -> u8 {
+        unsafe { GDALGetDataTypeSizeBits(self.gdal_type()) }
+            .try_into()
+            .unwrap()
+    }
+
+    /// Get the [`GDALDataType`] size in **bytes**.
+    pub fn bytes(&self) -> u8 {
+        unsafe { GDALGetDataTypeSizeBytes(self.gdal_type()) }
+            .try_into()
+            .unwrap()
+    }
+
+    /// Returns `true` if [`GDALDataType`] is integral (non-floating point)
+    pub fn is_integer(&self) -> bool {
+        (unsafe { GDALDataTypeIsInteger(self.gdal_type()) }) > 0
+    }
+
+    /// Returns `true` if [`GDALDataType`] is floating point (non-integral)
+    pub fn is_floating(&self) -> bool {
+        (unsafe { GDALDataTypeIsFloating(self.gdal_type()) }) > 0
+    }
+
+    /// Returns `true` if [`GDALDataType`] supports negative values.
+    pub fn is_signed(&self) -> bool {
+        (unsafe { GDALDataTypeIsSigned(self.gdal_type()) }) > 0
+    }
+
+    /// Return the descriptor for smallest [`GDALDataType`] fully contains both data types
+    /// indicated by `self` and `other`.
+    ///
+    /// # Example
+    ///
+    /// ```rust, no_run
+    /// use gdal::raster::GdalType;
+    /// println!("To safely store all possible '{}' and '{}' values, you should use  '{}'",
+    ///     <f32>::descriptor(),
+    ///     <i32>::descriptor(),
+    ///     <f32>::descriptor().union(<i32>::descriptor())
+    /// );
+    /// ```
+    pub fn union(&self, other: Self) -> Self {
+        let gdal_type = unsafe { GDALDataTypeUnion(self.gdal_type(), other.gdal_type()) };
+        Self(gdal_type)
+    }
+
+    /// Change a given value to fit within the constraints of this [`GDALDataType`].
+    ///
+    /// Returns an enum indicating if the wrapped value is unchanged, clamped
+    /// (to min or max datatype value) or rounded (for integral data types).
+    ///
+    /// # Example
+    ///
+    /// ```rust, no_run
+    /// use gdal::raster::{GdalType, AdjustedValue::*};
+    /// assert_eq!(<u8>::descriptor().adjust_value(255), Unchanged(255.));
+    /// assert_eq!(<u8>::descriptor().adjust_value(1.2334), Rounded(1.));
+    /// assert_eq!(<u8>::descriptor().adjust_value(1000.2334), Clamped(255.));
+    /// ```
+    pub fn adjust_value<N: GdalType + Into<f64>>(&self, value: N) -> AdjustedValue {
+        let mut is_clamped: libc::c_int = 0;
+        let mut is_rounded: libc::c_int = 0;
+
+        let result = unsafe {
+            GDALAdjustValueToDataType(
+                self.gdal_type(),
+                value.into(),
+                &mut is_clamped,
+                &mut is_rounded,
+            )
+        };
+
+        match (is_clamped > 0, is_rounded > 0) {
+            (false, false) => AdjustedValue::Unchanged(result),
+            (true, false) => AdjustedValue::Clamped(result),
+            (false, true) => AdjustedValue::Rounded(result),
+            (true, true) => panic!("Unexpected adjustment result: clamped and rounded."),
```

Comment:
Appears to be correct, although the GDAL code is somewhat subtle: for `GDT_Float32` it takes a `double`, casts it to `float`, then back to `double`  :+1:.

Reply:
🤦